sda-commons-server-jackson/src/main/java/org/sda/commons/server/jackson/hal/HalLinkProvider.java

```diff
@@ -0,0 +1,129 @@
+package org.sda.commons.server.jackson.hal;
+
+import io.openapitools.jackson.dataformat.hal.HALLink;
+import io.openapitools.jackson.dataformat.hal.HALLink.Builder;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.QueryParam;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.Feature;
+import javax.ws.rs.core.FeatureContext;
+import javax.ws.rs.core.UriBuilder;
+import javax.ws.rs.core.UriInfo;
+import org.glassfish.jersey.uri.internal.JerseyUriBuilder;
+import org.sda.commons.server.jackson.hal.HalLinkInvocationStateUtility.MethodInvocationState;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Helper utility that can be registered in the Jersey environment to be injected into services. It
+ * provides the context-based HALLink for a JAX-RS interface using the {@linkplain
+ * javax.ws.rs.PathParam} and {@linkplain javax.ws.rs.QueryParam} annotations.
+ *
+ * <p>Usage:
+ *
+ * <pre>
+ * &#064;Path("")
+ * interface TestApi {
+ *  &#064;Path("/testPath/{testArg}")
+ *  &#064;GET
+ *  String testMethod(@PathParam("testArg") String testArg, @QueryParam("query") query);
+ * }
+ *
+ * HALLink link = halLinkProvider.linkTo(methodOn(TestApi.class).testMethod("ResourceID", "detailed"));
+ * </pre>
+ *
+ * Would create a HALLink with the following URL: {@code
+ * "contextBasePath/testPath/ResourceID?query=detailed"}
+ */
+public class HalLinkProvider implements Feature {
+
+  private static final Logger LOG = LoggerFactory.getLogger(HalLinkProvider.class);
+
+  @Context private UriInfo uriInfo;
+
+  /**
+   * Creates a {@linkplain HALLink} based on JAX-RS annotated parameters of an interface. This
+   * method requires a second entrypoint with {@linkplain
+   * HalLinkInvocationStateUtility#methodOn(Class)} to process the annotated JAX-RS interface with
+   * the corresponding {@linkplain PathParam} and {@linkplain QueryParam} and the passed arguments
+   * to the proxied method invocation.
+   *
+   * <p>Example:
+   *
+   * <pre>
+   * &#064;Path("")
+   * interface TestApi {
+   *  &#064;Path("/testPath/{testArg}")
+   *  &#064;GET
+   *  String testMethod(@PathParam("testArg") String testArg, @QueryParam("query") query);
+   * }
+   *
+   * HALLink link = halLinkProvider.linkTo(methodOn(TestApi.class).testMethod("ResourceID", "detailed"));
+   * </pre>
+   *
+   * The example would create a HALLink with the following URL: {@code
+   * "contextBasePath/testPath/ResourceID?query=detailed"}
+   *
+   * @param invocation the invocation placeholder.
+   * @return the generated HALLink based on the method invocation
+   * @throws HalLinkMethodInvocationException - If no method invocation is provided via {@linkplain
+   *     HalLinkProvider#methodOn(Class)}
+   */
+  public HALLink linkTo(Object invocation) {
```

Comment:
Should we provide an API to get the URI as well? Something like

```java
URI uri = linkTo(methodOn(MyApi.class).getMyResource("123")).asUri();
HALLink link = linkTo(methodOn(MyApi.class).getMyResource("123")).asHalLink();
```

Reply:
Now we make use of the LinkResult
```
/** Wrapper class to provide the processed Link as {@linkplain URI} or {@linkplain HALLink} */
public final class LinkResult {
  private final URI uri;
  private final HALLink halLink;
…
```
Do you have that in mind?

Reply:
I just had the API in mind. The solution looks fine.